### ISSUE TITLE
Performance problems of the initial rendering of various editor modes

### ISSUE THREAD

**Issue by avoskresensky** [react-page/react-page#962]
**Describe the bug**
The performance of the initial rendering of various editor modes seems to significantly differ.
By initial rendering I mean the moment a user switches between modes using the toolbar.

According to our observations and some measurements, the winner, interestingly enough, is the "edit" mode, while the absolute loser is, interestingly enough, the "resize" mode.

We clearly notice the difference with some huge pages we have here in our internal app, which has a bunch of custom plugins. So to rule out the plugin interference, I'll try to demonstrate using a standard example page.

The measurements show that the difference in performance is not caused by how long it takes for the component tree to reconcile (render), but how long it takes to commit it, which seems to be a function of the number of state altering side effects.

The best way to demonstrate the problem is by using the Timings feature of the Performance tab of Chrome Dev Tools. It uses the Timing API which was removed from the latest versions of React, so in order to reproduce the attached measurements you'll need to downgrade your React to 16.x (I used 16.14).

In the below screenshots I'm using cellSpacing example and simply switching between editor modes.
It makes a difference which mode you are switching *from*, so I took all measurements switching from "preview" to the target mode. When measuring "preview" performance I was switching from "edit".

I have some Chrome plugins enabled, so absolute numbers are useless. However, when compared to each other they seem to make sense.
The reconciliation phase is about the same in all the measurements, but the commit phase differs greatly.

The "edit" mode is the only one that doesn't seem to have any state altering side effects. Here is the graph for it:

![image](https://user-images.githubusercontent.com/10179521/114810135-50072980-9d60-11eb-825d-c6ca3c951ed1.png)

"Insert" and "preview" modes seem to be about 50% slower, most likely because of some state altering side effect (cascading update warning):

![image](https://user-images.githubusercontent.com/10179521/114817144-54d2da00-9d6e-11eb-8a36-ed40519a0c6a.png)

The "move" mode is usually about 2 times slower, again with the cascading update warning:

![image](https://user-images.githubusercontent.com/10179521/114810671-6cf02c80-9d61-11eb-8c62-42fa2a74d2d1.png)

And here's our champ, the "resize" mode:

![image](https://user-images.githubusercontent.com/10179521/114815801-d07f5780-9d6b-11eb-9e8b-43ae50c4e0d0.png)

It's usually 10-15 times slower than "edit"!
It shows the same mammoth number regardless of whether you enter or exit it.
The more plugins you have on the page, the bigger hit the "resize" mode sluggishness is going to take. Here is what we see in our case on a huge page with lots of plugins:

![image](https://user-images.githubusercontent.com/10179521/114816454-0709a200-9d6d-11eb-8617-e01a8c49c7c9.png)

Another way to look at it is using a more standard React DevTools profiler. For 17+ React it is the only option. It lacks the cascading update warning, it doesn't clearly show you the time it took to perform the final stages of the commit, but you can tell the unhealthy cycle by the number of commits.

cellSpacing example page shows the difference of about 3x times more commits for "preview to resize" switch than "preview to edit" switch. This is a large number, but it's not that staggering. In our application, on a page with 120 plugins we have this:

Switching from "preview" to "edit":

![image](https://user-images.githubusercontent.com/10179521/114818316-73d26b80-9d70-11eb-825d-ecb288f25b77.png)

And here is switching from "preview" to "resize":

![image](https://user-images.githubusercontent.com/10179521/114818408-9a90a200-9d70-11eb-8c44-8c5cfb434bb8.png)

It looks like a bunch of state altering size effects somewhere in or around `ResizableRowCell` component.

**Expected behavior**
Resize mode clearly needs fixing, but it would be nice get rid of cascading update warnings across the board.

**Comment by macrozone** [comment]
thanks for the detailed analysis. 

is the worse performance of the resize mode a problem or is it just an observation?

is "resize in edit mode" enabled (its enabled by default)?

ResizableRowCell is always mounted (if not readonly), so its probably that `Draggable` fires an event very often.

What we also should look at is `useMeasure`, maybe it has to recalculate very often due to some reasons.

**Comment by zehawki** [comment]
Ah, glad you mentioned this - I was thinking this was my imagination, but I distinctly noticed a bit of _sluggishness_ when switching to edit from preview on a heavy duty page. This is after upgrading to v2.0.0 from v1.7.2.

**Comment by zehawki** [comment]
This is actually quite serious - this page (https://www.scoopcitizen.net/home) was taking 2 secs to move from 
1. preview to edit and 
2. revert from editing to normal

Now taking 5s for # 1 and 15 - 21 seconds for # 2. The browser tab is completely unresponsive during this duration.

Even allowing for react debug vs production mode, this is really long.

**Comment by avoskresensky** [comment]
@macrozone,

When it takes over a second to perform a commit, then I'm afraid it's a real problem. The plugins are all in their read-only state, so it has no right to be slower than "edit", especially given the fact that we have "resize in edit" on (which answers your other question).

Ok, I looked at useMeasure implementation. It very much looks like our culprit!
Sets the state every time the observer fires. Yikes!
Since those calls will be performed in "outside of react" context, all those `setState` calls will be unbatched -- each of them will cause their own commit which is exactly what I'm seeing. Gotta batch them somehow and apply all at once.

**Comment by avoskresensky** [comment]
@macrozone,

Okay, so we got three independent areas of concern. The last one is by far the most impacting.

* react-draggable causes instances of "cascading updates" and "forced reflow". These have little effect, but are trivial to fix. I'll file an issue with them.

* removing useMeasure does improve performance to an extent. It doesn't help the resize mode phenomenon, but it improves everything else. Right now it is used proactively and the associated observer etc is attached regardless of whether they even going to use the resize handles. The more cells there are, the costlier the whole thing becomes. It might be a good idea to consider enabling the associated logic only after they mouse down on the resize handle (or something like that).

* and finally this: the whole problem with sluggish resize mode switch goes away if this block is commented out inside Cell/Draggable/index.tsx:

```
if (isResizeMode) {
  return <>{children}</>;
}
```

This is the current performance graph with all code intact (cellSpacing plugin):
![image](https://user-images.githubusercontent.com/10179521/114965186-fa488500-9e24-11eb-8c53-daafd7031b70.png)

And this is the graph with the above block commented out:
![image](https://user-images.githubusercontent.com/10179521/114965410-5f9c7600-9e25-11eb-8712-c73bba3ed978.png)

It turns out that the problem is related to the fact the "resize" is the only mode that bails out of building the "long" version of the component tree, so React has to tear the old structure down, unmount everything, run a ton of associated lifecycles, them mount everything and run another bunch of the associated lifecycles. Here is switching from "edit" to "resize":

![image](https://user-images.githubusercontent.com/10179521/114967152-c3746e00-9e28-11eb-989e-1699bbc58d67.png)

Interestingly enough, even the "preview" mode keeps the long version of the tree.
Here's switching from "preview" to "edit": (notice all those "mounts" are "updates" now).

![image](https://user-images.githubusercontent.com/10179521/114967270-fd457480-9e28-11eb-8878-faa02fe4ed6c.png)

An interesting case of how making the component structure lean can lead to devastating results!
We trim the component structure and speed up to a negligible extent on rendering, but loose bit time on lifecycles.

**Comment by macrozone** [comment]
> It turns out that the problem is related to the fact the "resize" is the only mode that bails out of building the "long" version of the component tree, so React has to tear the old structure down, unmount everything, run a ton of associated lifecycles, them mount everything and run another bunch of the associated lifecycles. Here is switching from "edit" to "resize":

thats an interesting observation and makes kinda sense. So we just remove the 

```
 if (isResizeMode) {
    return <>{children}</>;
  }
```

?

**Comment by avoskresensky** [comment]
@macrozone,

I didn't make sure it works with this removed. I was just looking for an explanation for the phenomenon.

Some of the drag-n-drop logic that this early return circumvents might kick in and break that mode.
Some of it might need to be disabled.

**Comment by macrozone** [comment]
@avoskresensky just checked. It works when removed, but you can drag cells around in resize mode then (which might be actually useful).

there are some other early returns in the component there. I can try to streamline them a bit

**Comment by avoskresensky** [comment]
@macrozone,

> It works when removed, but you can drag cells around in resize mode then (which might be actually useful).

I would rather have resize handles in the layout mode, lol !

**Comment by macrozone** [comment]
> @macrozone,
> 
> > It works when removed, but you can drag cells around in resize mode then (which might be actually useful).
> 
> I would rather have resize handles in the layout mode, lol !

In earlier versions the modes were more separated. I think we could actually merge them and remove resize mode completly

**Comment by avoskresensky** [comment]
@macrozone,

> I think we could actually merge them and remove resize mode completly

I agree. It seems rather superfluous now. We found a good use for the layout mode however. We make those overlays that currently show on hover permanently visible via CSS and this mode turns into a sorta map of what's where:

![image](https://user-images.githubusercontent.com/10179521/115006487-cccbfd80-9e5d-11eb-9466-f1d434ab3e89.png)

Used some styling from v0.xxx to make those overlays contained.
The only thing we're missing there is resize handles from the resize/edit mode.

**Comment by macrozone** [comment]
we could also remove `allowMoveInEditMode` option and just always enable it. It was an experimental feature at some point, but i think there is little reason to disable it.

**Comment by avoskresensky** [comment]
No objections!

**Comment by macrozone** [comment]
> @macrozone,
> 
> > I think we could actually merge them and remove resize mode completly
> 
> I agree. It seems rather superfluous now. We found a good use for the layout mode however. We make those overlays that currently show on hover permanently visible via CSS and this mode turns into a sorta map of what's where:
> 
> ![image](https://user-images.githubusercontent.com/10179521/115006487-cccbfd80-9e5d-11eb-9466-f1d434ab3e89.png)
> 
> Used some styling from v0.xxx to make those overlays contained.
> The only thing we're missing there is resize handles from the resize/edit mode.

it actually used to look exactly like that. I move the labels up, so that it won't overlap the content. But it would be nice to have them inside the cells again at least in layout mode

**Comment by macrozone** [comment]
removing the options would be a breaking change though (strictly speaking). So i first clean it up a bit and fix the early returns, push it under v2

**Comment by macrozone** [comment]
@avoskresensky see https://github.com/react-page/react-page/pull/968, what do you think? (it does not remove and change any functionality)

**Comment by avoskresensky** [comment]
@macrozone ,

looking real solid. Tried it in our app and it works as expected.
Thanks!

**Comment by zehawki** [comment]
> 
> 
> This is actually quite serious - this page (https://www.scoopcitizen.net/home) was taking 2 secs to move from
> 
>     1. preview to edit and
> 
>     2. revert from editing to normal
> 
> 
> Now taking 5s for # 1 and 15 - 21 seconds for # 2. The browser tab is completely unresponsive during this duration.
> 
> Even allowing for react debug vs production mode, this is really long.

There's a remarkable difference on # 2. This is back to 2s. But # 1 is still 5secs or so.
